File: Leetcode/236-Lowest-Common-Ancestor-of-a-Binary-Tree/LowestCommonAncestorOfABinaryTree.py
# Python 3 solution

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        return self.lca(root, p, q);

    def lca(self, root, p, q):
        if root is None or p is None or q is None:
            return None
        if root == p or root == q:
            return root

        left = self.lca(root.left, p, q)
        right = self.lca(root.right, p, q)

        if left is not None and right is not None:
            return root

        return left if left is not None else right

File: Leetcode/236-Lowest-Common-Ancestor-of-a-Binary-Tree/LowestCommonAncestorOfABinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class LowestCommonAncestorOfABinaryTree {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || p == null || q == null) 
            return null;
        
        return lca(root, p, q);
    }
    
    private TreeNode lca(TreeNode root, TreeNode p, TreeNode q) {
        // if (root == null || root.val == p.val || root.val == q.val) 
        if (root == null || root == p || root == q)
            return root;
            
        TreeNode left = lca(root.left, p, q);
        TreeNode right = lca(root.right, p, q);  
        if (left!=null && right!=null) 
            return root;
        
        return left==null?right:left;
    }
}




